### PR TITLE
Turn on breakable prefill cuda graph for dp attention by default

### PR DESCRIPTION
**Pull request by merrymercy**
## Summary
- remove the DP-attention BCG validation blocker

## Testing
- python3 -m py_compile python/sglang/srt/server_args.py

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #29730567467](https://github.com/sgl-project/sglang/actions/runs/29730567467)
Latest PR Test (Extra): :white_check_mark: [Run #29771288654](https://github.com/sgl-project/sglang/actions/runs/29771288654)

### CODE REVIEW

File: test/registered/attention/unittests/dsv4/test_deepseek_v4.py [hunk 1]
```diff
@@ -395,6 +395,19 @@ def test_bcg_is_explicit_and_dsv4_backend_opt_in_only(self):
             DeepseekV4AttnBackend.use_captured_forward_metadata_for_breakable_cuda_graph
         )
 
+    def test_converted_extend_uses_prefill_metadata_mode(self):
```
Comment (merrymercy) [inline test/registered/attention/unittests/dsv4/test_deepseek_v4.py]: revert this useless test

File: python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py [hunk 2]
```diff
@@ -433,7 +436,12 @@ def _draft_extend_for_prefill(
                 forward_batch.spec_info.hidden_states = (
                     output.logits_output.hidden_states
                 )
-            if forward_batch.extend_seq_lens is not None:
+            # DP padding fabricates extend metadata on idle ranks, but their
```
Comment (merrymercy) [inline python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py]: is this still needed?

File: python/sglang/srt/model_executor/runner_backend/breakable_cuda_graph_backend.py [hunk 3]
```diff
@@ -137,6 +138,10 @@ def capture_one(
         stored = self._slice_output(self._shared_output_buffer, out_rows)
         self._graphs[shape_key] = graph
         self._outputs[shape_key] = stored
+        if dummies is not None:
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/runner_backend/breakable_cuda_graph_backend.py]: drop this if/else check, you can just do `self._outputs[shape_key] = dummies`

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 4]
```diff
@@ -879,7 +879,7 @@ def run_once():
         self.backend.capture_one(
             ShapeKey(size=num_tokens),
             run_once,
-            dummies=None,
+            dummies=forward_batch,
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: what is dummies?

File: python/sglang/srt/layers/attention/deepseek_v4_backend.py [hunk 5]
```diff
@@ -94,6 +94,12 @@ def _get_logical_forward_mode(forward_batch: ForwardBatch) -> ForwardMode:
     # from a reused/padded ForwardBatch turn an empty rank into TARGET_VERIFY.
     if forward_batch.forward_mode.is_idle():
         return forward_batch.forward_mode
+    # DP max-len padding can convert decode/idle/verify ranks into EXTEND so
+    # every rank runs the same prefill-shaped body. Plan metadata for the body
+    # that will actually execute; otherwise DSV4 misses prefill-only buffers
+    # such as c4_sparse_raw_indices.
+    if forward_batch.forward_mode.is_extend_without_speculative():
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/deepseek_v4_backend.py]: there is a pr just got merged. 
https://github.com/sgl-project/sglang/pull/31705

please rebase

File: test/registered/attention/unittests/dsv4/test_deepseek_v4.py (hunk 1 above)
Comment (merrymercy) [reply test/registered/attention/unittests/dsv4/test_deepseek_v4.py]: Removed. After rebasing onto #31705, the overlapping DSV4 implementation is already on main, so this redundant test and its standalone commit are no longer part of this PR.

File: python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py (hunk 2 above)
Comment (merrymercy) [reply python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py]: No. The per-step shallow copy keeps DP-fabricated extend metadata off the original idle ForwardBatch, where extend_seq_lens remains None. I removed the explicit idle check and kept the existing extend_seq_lens guard.

File: python/sglang/srt/model_executor/runner_backend/breakable_cuda_graph_backend.py (hunk 3 above)
Comment (merrymercy) [reply python/sglang/srt/model_executor/runner_backend/breakable_cuda_graph_backend.py]: Done: the capture input is now stored unconditionally. I kept it in a separate _capture_inputs map because _outputs[shape_key] holds the model output returned by replay; replacing that with the dummy ForwardBatch would break the replay contract.

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py (hunk 4 above)
Comment (merrymercy) [reply python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: It is the capture-time dummy ForwardBatch. DP padding may install newly allocated tensors on that batch, and BCG records their addresses, so the backend must retain the batch for the graph lifetime. I added this explanation at the call site and named the retained map _capture_inputs.

File: python/sglang/srt/layers/attention/deepseek_v4_backend.py (hunk 5 above)
Comment (merrymercy) [reply python/sglang/srt/layers/attention/deepseek_v4_backend.py]: Rebased onto latest main at 7d64858093, including #31705. I kept upstream's DSV4 implementation and removed this PR's overlapping DSV4 change and redundant test.

Comment (merrymercy) [reply python/sglang/srt/layers/attention/deepseek_v4_backend.py]: Follow-up: main advanced during the GPU test, so I rebased once more. The branch is now based on 1a317839d7 and pushed at d7e10eb64d.

**Comment by merrymercy** [comment]
/rerun-test test/registered/models_e2e/test_deepseek_v4_flash_fp4_b200.py test/registered/models_e2e/test_mimo_v2_flash.py

**Review by Oasis-Git — APPROVED** [review]

**Comment by merrymercy** [comment]
All CI is green.

### CI States

Latest PR Test (Base): :white_check_mark: [Run #29730567467](https://github.com/sgl-project/sglang/actions/runs/29730567467)
Latest PR Test (Extra): :white_check_mark: [Run #29771288654](https://github.com/sgl-project/sglang/actions/runs/29771288654)


merge now
